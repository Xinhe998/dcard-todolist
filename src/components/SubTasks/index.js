import React, { useContext, useRef } from 'react';
import PropTypes from 'prop-types';
import uuid from '../../utils';
import Store from '../../reducers/context';
import * as action from '../../actions/todoActions';
import TodoItem from '../TodoItem';

const SubTasks = ({ item }) => {
  const dispatch = useContext(Store);
  const shouldFocusRef = useRef();
  const focusWhenAdd = () => {
    setTimeout(() => {
      if (shouldFocusRef.current) shouldFocusRef.current.focus();
    }, 0);
  };
  return (
    <div className="CardScrollView__detail__block">
      <div className="CardScrollView__detail__block__title">SUB TASKS</div>
      {item.subtask && item.subtask.length > 0
        ? item.subtask.map((task, index) => (
          <TodoItem
            key={task.id}
            text={task.text}
            isComplete={task.isComplete}
            allowEdit
            inputRef={
              item.subtask.length === index + 1 ? shouldFocusRef : null
            }
            onClickCheckbox={() => {
              dispatch(
                action.updateSubTask({
                  id: item.id,
                  subtask: {
                    id: task.id,
                    isComplete: !task.isComplete,
                  },
                }),
              );
            }}
            del={() => {
              dispatch(action.deleteTodoItem(task));
            }}
            EditInputPlaceholder="Add a new subtask"
            handleEdit={e => dispatch(
              action.updateSubTask({
                id: item.id,
                subtask: {
                  id: task.id,
                  text: e.target.value,
                  isComplete: false,
                },
              }),
            )
            }
            handleBlur={(e) => {
              if (!e.target.value) {
                dispatch(action.deleteSubTask({
                  id: item.id,
                  subtask: {
                    id: task.id,
                  },
                }));
              }
            }}
          />
        ))
        : null}
      <TodoItem
        key={item.id}
        allowEdit
        EditInputPlaceholder="Add a new subtask"
        handleFocus={(e) => {
          dispatch(
            action.addSubTask({
              id: item.id,
              subtask: {
                id: uuid(),
                text: e.target.value,
                isComplete: false,
              },
            }),
          );
          focusWhenAdd();
        }}
        
      />
    </div>
  );
};

SubTasks.propTypes = {
  item: PropTypes.object,
};

SubTasks.defaultProps = {
  item: null,
};
export default SubTasks;
