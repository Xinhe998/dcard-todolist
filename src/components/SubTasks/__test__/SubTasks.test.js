/* eslint-disable func-names */
import React from 'react';
import { configure, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SubTasks from '../index';

import uuid from '../../../utils/uuid';

// 以該解析器提供給 Enzyme 做渲染 Component 的設置
configure({ adapter: new Adapter() });

describe('<SubTasks />', function () {
  let testComp;
  beforeEach(() => {
    jest.resetModules();
    this.params = {
      item: [],
    };

    this.makeSubject = () => {
      const { item } = this.params;

      testComp = mount(<SubTasks item={item} />);
      return testComp;
    };
  });

  describe('when it has complete data', () => {
    beforeEach(() => {
      // 給兩筆資料
      this.params.item = [
        { id: uuid(), text: 'test1', isComplete: false },
        { id: uuid(), text: 'test2', isComplete: true },
      ];
      this.subject = this.makeSubject();
    });

    it('should be same as snapshot', () => {
      const wrapper = render(testComp);
      expect(wrapper).toMatchSnapshot();
    });
    // 測試會render兩個TodoItem
    it('should render two <TodoItem>', () => {
      expect(testComp.find('.CardScrollView__item').children()).toHaveLength(2);
    });
  });
});
