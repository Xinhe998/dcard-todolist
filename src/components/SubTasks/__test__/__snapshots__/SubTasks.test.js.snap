// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubTasks /> when it has complete data should be same as snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "CardScrollView__detail__block",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "CardScrollView__detail__block__title",
        },
        "children": Array [
          Object {
            "data": "SUB TASKS",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "CardScrollView__item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "type": "checkbox",
                "value": "false",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "editInput",
                  "placeholder": "Add a new subtask",
                  "type": "text",
                  "value": "",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "type": undefined,
                "value": undefined,
              },
            },
            Object {
              "attribs": Object {
                "class": "editInput",
                "placeholder": "Add a new subtask",
                "type": "text",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "type": "checkbox",
                  "value": "false",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "type": undefined,
                  "value": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "CardScrollView__item",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "type": "checkbox",
              "value": "false",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "editInput",
                "placeholder": "Add a new subtask",
                "type": "text",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "type": undefined,
              "value": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "editInput",
              "placeholder": "Add a new subtask",
              "type": "text",
              "value": "",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "type": "checkbox",
                "value": "false",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "type": undefined,
                "value": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "CardScrollView__detail__block__title",
          },
          "children": Array [
            Object {
              "data": "SUB TASKS",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
